In [16]:
from GradientGang.Pipeline.FinalPipeline import FinalPipeline

## Initialize FinalPipeline

In [17]:
data_params = {
    'data_dir': "../dataset/PirateProcessed/",
    'train_file_name': "pirate_pain_train.csv",
    'train_file_name_labels': "pirate_pain_train_labels.csv",
    'test_file_name': "pirate_pain_test.csv",
    'train_global_features_file': "train_global_features.csv",
    'test_global_features_file': "test_global_features.csv",
    'batch_size': 128,
    'num_workers': 0,
    'val_split': 0.2,
    'shuffle': True,
    'use_kfold': True,
    'n_folds': 5,
}

In [18]:
# Initialize FinalPipeline with parameters
import os
import dotenv
env_path = '../src/GradientGang/Pipeline/Optimizer/.env'
dotenv.load_dotenv(env_path)
database_url = os.getenv("DATABASE_URL")

In [19]:
params = {
    "database_url": database_url,
    "data_params": data_params,
    'project_name': 'pirate_pain_classification',
    'study_name': 'final_kaggle_LAST',
    'submission_path': '../Submissions',
}

pipeline = FinalPipeline(params)

Storage: postgresql://postgres...
✓ Database configuration loaded
✓ Database initialized successfully!


[I 2025-11-17 12:36:14,710] Using an existing study with name 'pirate_pain_classificationfinal_kaggle_LAST' instead of creating a new one.


✓ Study created/loaded successfully!
Study name: pirate_pain_classificationfinal_kaggle_LAST
Sampler: TPESampler
Pruner: PatientPruner
Storage: Database
Total trials: 18
✓ Study initialized successfully!
✓ FinalPipeline initialized successfully!


## RUN FinalPipeline

In [20]:
#pipeline.optuna_optimize(n_trials=100000000000)  # Start the optimization process

## Track FinalPipeline

In [21]:
pipeline.study_summary()  # Display study summary

Study name: pirate_pain_classificationfinal_kaggle_LAST
Direction: 2
Total trials: 18
Completed trials: 16
Failed trials: 0
Pruned trials: 1
Running trials: 1

✓ Best trial: 0
✓ Best F1 score: 0.9598

Top 5 trials:
  1. Trial 0: F1=0.9598 | Autoencoder | MultiScaleCNN
  2. Trial 12: F1=0.9585 | Autoencoder | MultiScaleCNN
  3. Trial 11: F1=0.9559 | Autoencoder | MultiScaleCNN
  4. Trial 6: F1=0.9558 | Autoencoder | MultiScaleCNN
  5. Trial 4: F1=0.9542 | Autoencoder | MultiScaleCNN

📊 Trial states (last 10):
  ✓ Trial 8: COMPLETE | F1=0.9496
  ✓ Trial 9: COMPLETE | F1=0.9394
  ✓ Trial 10: COMPLETE | F1=0.9457
  ✓ Trial 11: COMPLETE | F1=0.9559
  ✓ Trial 12: COMPLETE | F1=0.9585
  ✓ Trial 13: COMPLETE | F1=0.9488
  ✓ Trial 14: COMPLETE | F1=0.9483
  ⊗ Trial 15: PRUNED | F1=0.9504
  ✓ Trial 16: COMPLETE | F1=0.9505
  ⟳ Trial 17: RUNNING | N/A


## Submit FinalPipeline

In [22]:
pipeline.create_submission()  # Generate submission file

Loading K-fold ensemble...

LOADING K-FOLD ENSEMBLE
Best trial: 0
Best mean F1: 0.9598
{'use_global_features': False, 'window_size': 31, 'jitter_strength': 0.0396879333044437, 'scaling_range': 0.0909213048309706, 'time_warp_strength': 0.0474386848806886, 'architectureType': 'MultiScaleCNN', 'numMultiScaleLayers': 2, 'kernelSizesIndex': 2, 'branchChannels_0': 33, 'branchChannels_1': 126, 'numFFLayers': 4, 'ffSizeCoefficient': 0.406169555339138, 'ffDropout': 0.0727299868828403, 'ReconstructionLossWeight': 0.0204926801154174, 'RegularizationWeight': 0.016480446427979, 'useClassWeights': True}
Architecture: Autoencoder
------------------------------------------------------------
Reconstructing architecture...
✓ Architecture reconstructed

Ensemble details:
  Fold 0: Expected F1 Score = 0.9753
  Fold 1: Expected F1 Score = 0.9589
  Fold 2: Expected F1 Score = 0.9458
  Fold 3: Expected F1 Score = 0.9656
  Fold 4: Expected F1 Score = 0.9534
----------------------------------------------------

GPU available: False, used: False
TPU available: False, using: 0 TPU cores


   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...


GPU available: False, used: False
TPU available: False, using: 0 TPU cores


   Training completed. Best F1: 0.9632
✓ Fold 0 retrained successfully: F1=0.9632
Fold 1: Checkpoint not found - retraining...

🔄 Retraining fold 1...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...


GPU available: False, used: False
TPU available: False, using: 0 TPU cores


   Training completed. Best F1: 0.9418
✓ Fold 1 retrained successfully: F1=0.9418
Fold 2: Checkpoint not found - retraining...

🔄 Retraining fold 2...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...


GPU available: False, used: False
TPU available: False, using: 0 TPU cores


   Training completed. Best F1: 0.9270
✓ Fold 2 retrained successfully: F1=0.9270
Fold 3: Checkpoint not found - retraining...

🔄 Retraining fold 3...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...


GPU available: False, used: False
TPU available: False, using: 0 TPU cores


   Training completed. Best F1: 0.9451
✓ Fold 3 retrained successfully: F1=0.9451
Fold 4: Checkpoint not found - retraining...

🔄 Retraining fold 4...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...
   Training completed. Best F1: 0.9359
✓ Fold 4 retrained successfully: F1=0.9359
------------------------------------------------------------
[OK] All 5 models loaded/retrained
Fold F1 scores: ['0.9632', '0.9418', '0.9270', '0.9451', '0.9359']
Average F1 score: 0.9426

✓ Ensemble created with performance-based weights
Weights: ['0.204', '0.200', '0.197', '0.201', '0.199']
Weighted F1 Estimate: 0.9427
Device: cpu
✓ Submission generated: ../Submissions\submission_20251117_124906.csv
Total predictions: 1324

Generating confusion matrix on validation set...
  Computing cross-validation predictions on full training set...
  Processing fold 1/5...
  Processing fold 2/5...
  Processin

,sample_index,label
0,000,no_pain
1,001,no_pain
2,002,no_pain
3,003,no_pain
4,004,no_pain
...,...,...
1319,1319,no_pain
1320,1320,no_pain
1321,1321,no_pain
1322,1322,no_pain


Loading K-fold ensemble...

Best trial: 0
Best mean F1: 0.9598
{'use_global_features': False, 'window_size': 31, 'jitter_strength': 0.0396879333044437, 'scaling_range': 0.0909213048309706, 'time_warp_strength': 0.0474386848806886, 'architectureType': 'MultiScaleCNN', 'numMultiScaleLayers': 2, 'kernelSizesIndex': 2, 'branchChannels_0': 33, 'branchChannels_1': 126, 'numFFLayers': 4, 'ffSizeCoefficient': 0.406169555339138, 'ffDropout': 0.0727299868828403, 'ReconstructionLossWeight': 0.0204926801154174, 'RegularizationWeight': 0.016480446427979, 'useClassWeights': True}
Architecture: Autoencoder


Reconstructing architecture...
✓ Architecture reconstructed

Ensemble details:
  Fold 0: Expected F1 Score = 0.9753
  Fold 1: Expected F1 Score = 0.9589
  Fold 2: Expected F1 Score = 0.9458
  Fold 3: Expected F1 Score = 0.9656
  Fold 4: Expected F1 Score = 0.9534
  


[STEP 1/3] Searching for checkpoints...


  Fold 0: Not found
  Fold 1: Not found
  Fold 2: Not found
  Fold 3: Not found
  Fold 4: Not found

Checkpoints found: 0/5

[STEP 2/3] Skipping consistency check (no checkpoints found)

[STEP 3/3] Loading/retraining 5 fold models...


Fold 0: Checkpoint not found - retraining...

🔄 Retraining fold 0...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
   Training completed. Best F1: 0.9632
✓ Fold 0 retrained successfully: F1=0.9632
Fold 1: Checkpoint not found - retraining...

🔄 Retraining fold 1...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
   Training completed. Best F1: 0.9418
✓ Fold 1 retrained successfully: F1=0.9418
Fold 2: Checkpoint not found - retraining...

🔄 Retraining fold 2...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
   Training completed. Best F1: 0.9270
✓ Fold 2 retrained successfully: F1=0.9270
Fold 3: Checkpoint not found - retraining...

🔄 Retraining fold 3...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
   Training completed. Best F1: 0.9451
✓ Fold 3 retrained successfully: F1=0.9451
Fold 4: Checkpoint not found - retraining...

🔄 Retraining fold 4...
   MacroArchitecture: Autoencoder
   Use windowing: True
   Setting up fold data...
   Train batches: 15, Val batches: 2
   Starting training (max 100 epochs)...
   Training completed. Best F1: 0.9359
✓ Fold 4 retrained successfully: F1=0.9359


[OK] All 5 models loaded/retrained
Fold F1 scores: ['0.9632', '0.9418', '0.9270', '0.9451', '0.9359']
Average F1 score: 0.9426

✓ Ensemble created with performance-based weights
Weights: ['0.204', '0.200', '0.197', '0.201', '0.199']
Weighted F1 Estimate: 0.9427
Device: cpu


✓ Submission generated: ../Submissions\submission_20251117_124906.csv
Total predictions: 1324

Generating confusion matrix on validation set...
  Computing cross-validation predictions on full training set...
  Processing fold 1/5...
  Processing fold 2/5...
  Processing fold 3/5...
  Processing fold 4/5...
  Processing fold 5/5...
  Total samples evaluated: 661
✓ Confusion matrix saved: ../Submissions\confusion_matrix_20251117_124906.png

Cross-Validation Metrics (Full Training Set):

Overall Accuracy: 0.9138

Per-Class Metrics:
    Class Precision Recall F1-Score  Support
  no_pain    0.9606 0.9530   0.9568      511
 low_pain    0.8681 0.8404   0.8541       94
high_pain    0.6032 0.6786   0.6387       56

✓ Metrics table saved: ../Submissions\cv_metrics_20251117_124906.csv
✓ Confusion matrix CSV saved: ../Submissions\confusion_matrix_20251117_124906.csv